### ISSUE TITLE
Token customizer not triggered on token exchange requests

### ISSUE THREAD

**Issue by fleskesvor** [spring-projects/spring-authorization-server#1692]
**Describe the bug**
I have a Spring authorization server set up with a token customizer (`OAuth2TokenCustomizer<JwtEncodingContext>`), and I'm able to customize tokens resulting from form login (`UsernamePasswordAuthenticationToken`) and OAuth2 login (`OAuth2AuthenticationToken`). 

Additionally, I've set up two applications, which both use this auth server for IAM, and have set up token exchange in one of them to call an endpoint in the other, based on the impersonation sample in [UserController.java](https://github.com/spring-projects/spring-authorization-server/blob/main/samples/users-resource/src/main/java/sample/web/UserController.java). Both applications verify the audience of tokens, and I can verify that this is working as expected.

However, when trying to customize `OAuth2TokenExchangeAuthenticationToken`s, as described in the [token exchange announcement](https://spring.io/blog/2024/03/19/token-exchange-support-in-spring-security-6-3-0-m3#enabling-token-exchange-on-the-server), the customizer is never triggered.

**To Reproduce**
Unfortunately, I can't share any of my original code, but I've set up a minimal authorization server with support for token exchange based on the documentation and aforementioned blog post:

- https://docs.spring.io/spring-authorization-server/reference/getting-started.html
- https://spring.io/blog/2024/03/19/token-exchange-support-in-spring-security-6-3-0-m3#enabling-token-exchange-on-the-server

I've tweaked the configuration slightly to enabled authorization code flow with a public client from a minimal single page application, since this is what I'm most familiar with, and I use that to fetch an access token with form login using the example user from the blog post.

Once I have that token, I use curl to trigger token exchange in my auth server:

```
curl -u token-client:token http://localhost:8080/oauth2/token
  -d 'grant_type=urn:ietf:params:oauth:grant-type:token-exchange'
  -d 'scope=message.read'
  -d 'subject_token_type=urn:ietf:params:oauth:token-type:jwt'
  -d 'subject_token=[TOKEN GOES HERE]'
```

This gives me a new access token with "token-client" as audience, and the requested scope as scope. However, the token customizer is only triggered on the `UsernamePasswordAuthenticationToken`, which doesn't hold any information about requested audience, resource, etc.

**Expected behavior**
`OAuth2TokenCustomizer` should be triggered with the `OAuth2TokenExchangeAuthenticationToken`, as described in the linked blog post, so that claims can be customized.

**Sample**

https://github.com/fleskesvor/auth-server-java
https://github.com/fleskesvor/demo-client - this is only to trigger form login and debug the resulting token.

**Comment by jgrandja** [comment]
@fleskesvor The blog post provided code sample for the `6.3.0-M3` milestone release, which is not stable at that point. The implementation changed by the time it was released in `6.3.0` GA, which made the code sample for `accessTokenCustomizer()` outdated and no longer correct.

Instead of:

```java
	@Bean
	public OAuth2TokenCustomizer<JwtEncodingContext> accessTokenCustomizer() {
		return (context) -> {
			if (context.getPrincipal() instanceof OAuth2TokenExchangeAuthenticationToken tokenExchangeRequest) {
				var resources = tokenExchangeRequest.getResources();
				// TODO: Validate resource value(s) and map to the
				// appropriate audience value(s) if needed...

				context.getClaims().audience(...);
			}
		};
	}
```

It should be:

```java
	@Bean
	public OAuth2TokenCustomizer<JwtEncodingContext> accessTokenCustomizer() {
		return (context) -> {
			if (AuthorizationGrantType.TOKEN_EXCHANGE.equals(context.getAuthorizationGrantType())) {
				OAuth2TokenExchangeAuthenticationToken tokenExchangeAuthentication = context.getAuthorizationGrant();
				Set<String> resources = tokenExchangeAuthentication.getResources();
				// TODO: Validate resource value(s) and map to the
				//  appropriate audience value(s) if needed...

				context.getClaims().audience(...);
			}
		};
	}	
```

In the future, I would recommend using code from a working sample, which you can find in the [Demo Sample](https://github.com/spring-projects/spring-authorization-server/tree/main/samples#demo-sample).

FYI, I updated the blog post with the correct sample code.

**Comment by fleskesvor** [comment]
Thanks for the reply; that works perfectly! :+1: I've actually used the samples extensively while implementing my authorization server (including for setting up a working implementation of token exchange between two applications), but I couldn't find an example of customizing token exchange tokens specifically. The tokens would just end up being handled by my generic customizer for access token, but I didn't think to look at the authorization grant type. Thanks again for pointing me in the right direction, and apologies for the "noise". :slightly_smiling_face:
